File: chapter12/code12_03.cpp
#include <iostream>
#include <vector>

using namespace std;

void QuickSort(vector<int> &a, int l, int r)
{
    if (r - l <= 1)
        return;

    int pivot_idx = (l + r) / 2;
    int pivot = a[pivot_idx];
    swap(a[pivot_idx], a[r - 1]);

    int i = l;
    for (int j = l; j < r - 1; ++j)
    {
        if (a[j] < pivot)
        {
            swap(a[i++], a[j]);
        }
    }
    swap(a[i], a[r - 1]);

    QuickSort(a, l, i);
    QuickSort(a, i + 1, r);
}

int main()
{
    int N;
    cin >> N;
    vector<int> a(N);
    for (int i = 0; i < N; ++i)
        cin >> a[i];
    QuickSort(a, 0, N);
    for (int i = 0; i < a.size(); ++i)
    {
        cout << a[i] << ": ";
    }
    cout << endl;
}

File: chapter12/code12_03.py
def quick_sort(L, l, r):
    if r-l > 1:
        pivot_idx = (l+r)//2
        pivot = L[pivot_idx]
        L[pivot_idx], L[r-1] = L[r-1], L[pivot_idx]

        i = l
        for j in range(l, r-1):
            if L[j] < pivot:
                L[i], L[j] = L[j], L[i]
                i += 1
        L[i], L[r-1] = L[r-1], L[i]
        quick_sort(L, l, i)
        quick_sort(L, i+1, r)


n = int(input())

L = list(map(int, input().split()))

quick_sort(L, 0, n)

print(L)
